### ISSUE TITLE
"no-member" false positive in inherited dataclasses

### ISSUE THREAD

**Issue by ppentchev** [PyCQA/pylint#7325]
### Bug description

[pyclass.txt](https://github.com/PyCQA/pylint/files/9387833/pyclass.txt)
Hi,

Thanks a lot for working on pylint, astroid, and the related projects!

It seems that pylint does not "listen to" isinstance() checks and type annotations when a program attempts to make sure that a variable that is guaranteed to be an instance of a base class is actually an instance of a specific class derived from the base one. See the attached file; pylint 2.14.5 / astroid 2.11.7 issue "no-member" messages for both the `print(expr.name)` and `print(expr.op_name)` lines.

Thanks in advance for your time, and keep up the great work!

### Command used

```shell
pylint pyclass.py
```

### Pylint output

```shell
************* Module pyclass
pyclass.py:86:10: E1101: Instance of 'Expr' has no 'name' member (no-member)
pyclass.py:97:10: E1101: Instance of 'Expr' has no 'op_name' member (no-member)

------------------------------------------------------------------
Your code has been rated at 7.83/10 (previous run: 7.73/10, +0.1
```

### Expected behavior

No diagnostic messages at all.

### Pylint version

```shell
pylint 2.14.5
astroid 2.11.7
Python 3.10.6 (main, Aug 10 2022, 11:19:32) [GCC 12.1.0]
```

### OS / Environment

Debian testing on a x86_64 host, Python from the python3 package (3.10.5-3), pylint installed via tox with a `deps: pylint >= 2.14` testenv configuration.

**Comment by jacobtylerwalls** [comment]
Duplicate of #4693 (and in turn #1162)
